crates/store/src/db/models/queries.rs

```diff
@@ -0,0 +1,1647 @@
+/// # Note
+///
+/// This method returns notes from a single block. To fetch all notes up to the chain tip,
+/// multiple requests are necessary.
+pub(crate) fn select_notes_since_block_by_tag_and_sender(
+    conn: &mut SqliteConnection,
+    start_block_number: BlockNumber,
+    account_ids: &[AccountId],
+    note_tags: &[u32],
+) -> Result<Vec<NoteSyncRecord>, DatabaseError> {
+    QueryParamAccountIdLimit::check(account_ids.len())?;
+    QueryParamNoteTagLimit::check(note_tags.len())?;
+    // SELECT
+    //     block_num,
+    //     batch_index,
+    //     note_index,
+    //     note_id,
+    //     note_type,
+    //     sender,
+    //     tag,
+    //     aux,
+    //     execution_hint,
+    //     inclusion_path
+    // FROM
+    //     notes
+    // WHERE
+    //     -- find the next block which contains at least one note with a matching tag or sender
+    //     block_num = (
+    //         SELECT
+    //             block_num
+    //         FROM
+    //             notes
+    //         WHERE
+    //             (tag IN rarray(?1) OR sender IN rarray(?2)) AND
+    //             block_num > ?3
+    //         ORDER BY
+    //             block_num ASC
+    //     LIMIT 1) AND
+    //     -- filter the block's notes and return only the ones matching the requested tags or
+    // senders     (tag IN rarray(?1) OR sender IN rarray(?2))
+
+    let desired_note_tags = Vec::from_iter(note_tags.iter().map(|tag| *tag as i32));
+    let desired_senders = serialize_vec(account_ids.iter());
+
+    let start_block_num = block_number_to_raw_sql(start_block_number);
+
+    // find block_num: select notes since block by tag and sender
+    let Some(desired_block_num): Option<i64> =
+        SelectDsl::select(schema::notes::table, schema::notes::block_num)
+            .filter(
+                schema::notes::tag
+                    .eq_any(&desired_note_tags[..])
+                    .or(schema::notes::sender.eq_any(&desired_senders[..])),
+            )
+            .filter(schema::notes::block_num.gt(start_block_num))
+            .order_by(schema::notes::block_num.asc())
+            .limit(1)
+            .get_result(conn)
+            .optional()?
+    else {
+        return Ok(Vec::new());
+    };
+
+    let notes = SelectDsl::select(schema::notes::table, NoteSyncRecordRawRow::as_select())
+            // find the next block which contains at least one note with a matching tag or sender
+            .filter(schema::notes::block_num.eq(
+                &desired_block_num
+            ))
+            // filter the block's notes and return only the ones matching the requested tags or senders
+            .filter(
+                schema::notes::tag
+                .eq_any(&desired_note_tags)
+                .or(
+                    schema::notes::sender
+                    .eq_any(&desired_senders)
+                )
+            )
+            .get_results::<NoteSyncRecordRawRow>(conn)
+            .map_err(DatabaseError::from)?;
```

Comment:
This sends two requests to the database while previously we sent only one request, right? This is because we cannot use `rarray` with `disel`, correct?

Probably not a big deal for now (and we may end up refactoring this anyway).

Reply:
Since sqlite is embedded there isn't much difference between performing the operation within the database, or within rust code. Usually the difference is

- database query plan optimiser might have additional information to optimise/index better than the rust code, and
- you can perform more complex or otherwise hidden from the db schema indexing/lookups on the rust end

But as you say, not so important right now